server/src/handler_driver.c

```diff
@@ -1275,46 +1396,61 @@ onion_connection_status handler_set_train_peripheral(void *_,
 		int client_session_id = params_check_session_id(data_session_id);
 		int grab_id = params_check_grab_id(data_grab_id, TRAIN_ENGINE_INSTANCE_COUNT_MAX);
 		int state = params_check_state(data_state);
+		
 		if (client_session_id != session_id) {
 			syslog_server(LOG_ERR, "Request: Set train peripheral - invalid session id");
 			return OCS_NOT_IMPLEMENTED;
+		} 
+		
+		pthread_mutex_lock(&grabbed_trains_mutex);
+		if (grab_id == -1 || !grabbed_trains[grab_id].is_valid) {
+			pthread_mutex_unlock(&grabbed_trains_mutex);
+			syslog_server(LOG_ERR, "Request: Set train peripheral - invalid grab id");
+			return OCS_NOT_IMPLEMENTED;
 		} else if (state == -1) {
-			syslog_server(LOG_ERR, "Request: Set train peripheral - bad state");
+			syslog_server(LOG_ERR, 
+			              "Request: Set train peripheral - train: %s - bad state", 
+			              grabbed_trains[grab_id].name->str);
+			pthread_mutex_unlock(&grabbed_trains_mutex);
 			return OCS_NOT_IMPLEMENTED;
 		} else if (data_peripheral == NULL) {
-			syslog_server(LOG_ERR, "Request: Set train peripheral - bad peripheral");
+			syslog_server(LOG_ERR, 
+			              "Request: Set train peripheral - train: %s - bad peripheral",
+			              grabbed_trains[grab_id].name->str);
+			pthread_mutex_unlock(&grabbed_trains_mutex);
 			return OCS_NOT_IMPLEMENTED;
 		} else if (data_track_output == NULL) {
-			syslog_server(LOG_ERR, "Request: Set train peripheral - bad track output");
+			syslog_server(LOG_ERR, 
+			              "Request: Set train peripheral - train: %s peripheral: %s - bad track output", 
+			              grabbed_trains[grab_id].name->str, data_peripheral);
+			pthread_mutex_unlock(&grabbed_trains_mutex);
 			return OCS_NOT_IMPLEMENTED;
 		}
 		
-		pthread_mutex_lock(&grabbed_trains_mutex);
-		if (grab_id == -1 || !grabbed_trains[grab_id].is_valid) {
+		syslog_server(LOG_NOTICE, 
+		              "Request: Set train peripheral - train: %s peripheral: %s state: 0x%02x", 
+		              grabbed_trains[grab_id].name->str, data_peripheral, state);
+		if (bidib_set_train_peripheral(grabbed_trains[grab_id].name->str,
+		                               data_peripheral, state,
+		                               data_track_output)) {
+			syslog_server(LOG_ERR, 
+			              "Request: Set train peripheral - train: %s "
+			              "peripheral: %s state: 0x%02x - bad parameter values",
```

Comment:
invalid parameters?

Reply:
Add track output to log message?

Reply:
track output not interesting IMO, same as in the earlier comments